### PR TITLE
[lldb][libc++] Hide all libc++ implementation details from stacktraces

### PR DESCRIPTION
**Pull request by vogelsgesang**
This commit changes the libc++ frame recognizer to hide implementation details of libc++ more aggressively. The applied heuristic is rather straightforward: We consider every function name starting with `__` as an implementation detail.

This works pretty neatly for `std::invoke`, `std::function`, `std::sort`, `std::map::emplace` and many others. Also, this should align quite nicely with libc++'s general coding convention of using the `__` for their implementation details, thereby keeping the future maintenance effort low.

However, this heuristic by itself does not work in 100% of the cases: E.g., `std::ranges::sort` is not a function, but an object with an overloaded `operator()`, which means that there is no actual call `std::ranges::sort` in the call stack. Instead, there is a `std::ranges::__sort::operator()` call. To make sure that we don't hide this stack frame, we never hide the frame which represents the entry point from user code into libc++ code

### CODE REVIEW

**Review by ldionne — COMMENTED** [review]
This feels like a good idea, but I am also a bit scared of hiding too much stuff and making it difficult to figure out what's going wrong from a stack trace. It's not rational, I have no evidence to back this up and perhaps this is only fear of unknown.

Factually, the heuristic of using `__` makes sense to me since all implementation-detail functions start with `__`, and nothing in the public libc++ API starts with `__` (off the top of my head).

Overall, I think I like this patch but like I said I'm a bit scared of it being too aggressive.

**Comment by vogelsgesang** [comment]
> [...] like I said I'm a bit scared of it being too aggressive.

Yes, this is also my main concern here. E.g., for `std::ranges::sort`, it already turns out to be slightly too aggressive, as discussed in the commit message and the test case comment. The "`std::ranges::sort` calls an operator() of an internal struct type" pattern is already slightly problematic...

Do you have an idea how we could de-risk this change? E.g., would do you have a couple more scenarios in mind for which I should add test cases? Are there similar patterns to the `operator()` which I should also cover?

Do we have some libc++ developers which are using lldb top-of-tree for their day-to-day work on libc++ and who would make us aware in case this commit turns out to be too aggressive?

**Comment by Michael137** [comment]
When you're stopped in a frame and you have libc++ frames above you in the backtrace, you're either stopped in a callback (like `std::function`, `std::make_unique`, `std::sort`), or you're deliberately stepping into libc++ code. For the latter, you probably don't want to hide any of the implementation details (though the libc++ devs might be able to comment better on their workflow here). I would find it weird if I stepped into a series of function calls manually but the backtrace omitted my path down.

I wonder if a heuristic for the callback case could be: mark all but the first libc++ entry point as an implementation detail. Would this be even more aggressive? Not sure. Wdyt?

**Comment by ldionne** [comment]
> Do you have an idea how we could de-risk this change? E.g., would do you have a couple more scenarios in mind for which I should add test cases? Are there similar patterns to the `operator()` which I should also cover in test cases?

What I could imagine is to drive this via attributes instead of a heuristic. We could potentially mark implementation details of libc++ as such to control what the debugging experience is like at a much finer grain if we used attributes, but obviously this would also increase complexity in the code and it would be yet another thing that we have to slap on almost every declaration.

> Do we have some libc++ developers which are using lldb top-of-tree for their day-to-day work on libc++ and who would make us aware in case this commit turns out to be too aggressive?

I'm not aware of anyone. I just use `lldb` provided by the platform usually.

**Comment by jimingham** [comment]
Do we need to worry about whether this recognizer hides frames that might be useful to someone actually working on the libc++ they are debugging?  I would imagine in that case, you'd probably want to turn off this recognizer anyway, since you want to see the details.

Seems to me like this feature should be tuned to "users of libc++" not "developers of libc++".

We really should add a "disable" to the frame recognizers.  Once you delete them you can't really get the built-in ones back.  People who are working ON libc++ still spend a bunch of their time debugging "uses" of it, and likely would benefit from the implementation hiding, then every so often want to see the details.  We should make that easier.

**Comment by vogelsgesang** [comment]
> We really should add a "disable" to the frame recognizers

Do you mean something like #109219? (Test cases still missing; rest should be ready for review)

> Seems to me like this feature should be tuned to "users of libc++" not "developers of libc++".

Agree, I think that should be the goal here

> I wonder if a heuristic for the callback case could be: mark all but the first libc++ entry point as an implementation detail. Would this be even more aggressive? Also, the implementation of this might be a bit awkward. Not sure. Wdyt?

🤔 Might be a good idea... I will have to experiment a bit with this...

**Comment by vogelsgesang** [comment]
> What I could imagine is to drive this via attributes instead of a heuristic. We could potentially mark implementation details of libc++ as such to control what the debugging experience is like at a much finer grain if we used attributes, but obviously this would also increase complexity in the code and it would be yet another thing that we have to slap on almost every declaration.

Mayve we could also find some other convention. E.g. hiding the `std::*::__detail` namespace. (You could still move all those `__` methods into a different namespace, right? They are not part of the stable ABI, are they?). Using a namespace would have the benefit that

1. you don't need to annotate each individual method, you can simply use a scope
2. we don't need to add a new attribute to clang, dwarf etc.

**Comment by vogelsgesang** [comment]
> We really should add a "disable" to the frame recognizers

Merged #109219

> I wonder if a heuristic for the callback case could be: mark all but the first libc++ entry point as an implementation detail. Would this be even more aggressive? Also, the implementation of this might be a bit awkward. Not sure. Wdyt?

I combined your approach with the `__`-based heuristic. I am quite happy with the result. It's more robust now, as can be seen by the updated test expectations for `ranges::sort`: The top-level `std::__1::ranges::__sort::operator()` is no longer hidden

File: lldb/test/API/lang/cpp/libcxx-internals-recognizer/TestLibcxxInternalsRecognizer.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    def test_frame_recognizer(self):
+        """Test that implementation details of libc++ are hidden"""
+        self.build()
+        (target, process, thread, bkpt) = lldbutil.run_to_source_breakpoint(
+            self, "break here", lldb.SBFileSpec("main.cpp")
+        )
+
+        expected_parents = {
+            "sort_less(int, int)": ["::sort", "test_algorithms"],
+            # `std::ranges::sort` is implemented as an object of types `__sort`.
```
Comment (frederick-vs-ja) [inline lldb/test/API/lang/cpp/libcxx-internals-recognizer/TestLibcxxInternalsRecognizer.py:20]: There're a large number of such customization point objects (and niebloids, which will be respecified as CPOs soon, see [P3136R0](https://wg21.link/p3136r0)) since C++20. Should we invent some convention to recognize them uniformly?

Comment (vogelsgesang) [reply lldb/test/API/lang/cpp/libcxx-internals-recognizer/TestLibcxxInternalsRecognizer.py:20]: Not sure that would be necessary. With the current heuristic ("never hide a frame which was immediately called from user code"), the end result to the user is fine, despite the debug info containing the `__sort::operator()` function name

File: libcxx/docs/UserDocumentation.rst [hunk 2]
```diff
@@ -346,6 +346,31 @@ Third-party Integrations
...
+
+Note how the caller of ``my_comparator`` is shown as ``std::sort``. Looking at
+the frame numbers, we can see that frames #1 until #6 were hidden. Those frames
+represent internal implementation details such as ``__sort4`` and similar
+utility functions.
+
+To also show those implementation details, use ``thread backtrace -u``.
+Alternatively, to disable those compact backtraces, use ``frame recognizer list``
+and ``frame recognizer disable`` on the "libc++ frame recognizer".
+
```
Comment (Michael137) [inline libcxx/docs/UserDocumentation.rst:373]: While we're here, could we also mention `target.process.thread.step-avoid-regexp`? Which is an LLDB setting set to `^std` by default. This disables stepping into libc++ functions. Happy to also just do that in a separate PR.

File: lldb/source/Plugins/LanguageRuntime/CPlusPlus/CPPLanguageRuntime.cpp [hunk 3]
```diff
@@ -90,9 +79,26 @@ class LibCXXFrameRecognizer : public StackFrameRecognizer {
     if (!sc.function)
       return {};
 
+    // Check if we have a regex match
+    bool matches_regex = false;
     for (RegularExpression &r : m_hidden_regex)
```
Comment (Michael137) [inline lldb/source/Plugins/LanguageRuntime/CPlusPlus/CPPLanguageRuntime.cpp:84]: Add braces to the outer loop

File: lldb/source/Plugins/LanguageRuntime/CPlusPlus/CPPLanguageRuntime.cpp [hunk 4]
```diff
@@ -90,9 +79,26 @@ class LibCXXFrameRecognizer : public StackFrameRecognizer {
...
-      if (r.Execute(sc.function->GetNameNoArguments()))
+      if (r.Execute(sc.function->GetNameNoArguments())) {
+        matches_regex = true;
+        break;
+      }
+
+    if (matches_regex) {
+      // Only hide this frame if the immediate caller is also within libc++.
+      lldb::StackFrameSP parent_frame =
```
Comment (Michael137) [inline lldb/source/Plugins/LanguageRuntime/CPlusPlus/CPPLanguageRuntime.cpp:92]: `parent_frame` -> `parent_frame_sp`.

Also, we need to check `parent_frame` and `GetThread()` for `nullptr` here

File: lldb/source/Plugins/LanguageRuntime/CPlusPlus/CPPLanguageRuntime.cpp [hunk 5]
```diff
@@ -90,9 +79,26 @@ class LibCXXFrameRecognizer : public StackFrameRecognizer {
...
+          frame_sp->GetThread()->GetStackFrameAtIndex(
+              frame_sp->GetFrameIndex() + 1);
+      const auto &parent_sc =
+          parent_frame->GetSymbolContext(lldb::eSymbolContextFunction);
+      if (parent_sc.function->GetNameNoArguments().GetStringRef().starts_with(
```
Comment (Michael137) [inline lldb/source/Plugins/LanguageRuntime/CPlusPlus/CPPLanguageRuntime.cpp:97]: `parent_sc.function` may not exist.

File: lldb/source/Plugins/LanguageRuntime/CPlusPlus/CPPLanguageRuntime.cpp [hunk 6]
```diff
@@ -90,9 +79,26 @@ class LibCXXFrameRecognizer : public StackFrameRecognizer {
...
+    // Check if we have a regex match
+    bool matches_regex = false;
     for (RegularExpression &r : m_hidden_regex)
-      if (r.Execute(sc.function->GetNameNoArguments()))
+      if (r.Execute(sc.function->GetNameNoArguments())) {
+        matches_regex = true;
+        break;
+      }
+
+    if (matches_regex) {
```
Comment (Michael137) [inline lldb/source/Plugins/LanguageRuntime/CPlusPlus/CPPLanguageRuntime.cpp:90]: Why split this out of the loop above?

Comment (vogelsgesang) [reply lldb/source/Plugins/LanguageRuntime/CPlusPlus/CPPLanguageRuntime.cpp:90]: purely stylistic. I find the code easier to read if every section takes care of a single thing.

In this case, I find it more readable if we have two separate code blocks:
1. first the code block that check if we match the regular expression
2. then a separate code block which checks for the parent frame

But this is clearly a subjective argument. If you want me to write this in some other way, please let me know 🙂

**Comment by vogelsgesang** [comment]
@Michael137 thanks for the review! I updated the pull request accordingly

File: libcxx/docs/UserDocumentation.rst (hunk 2 above)
Comment (vogelsgesang) [reply libcxx/docs/UserDocumentation.rst:373]: I assume that this mechanism can be disabled via `settings set target.process.thread.step-avoid-regexp ""`? Or would that match *every* frame and break stepping completely?

**Comment by vogelsgesang** [comment]
@jimingham I agree with everything you wrote. I think you misunderstood the context based on the way GitHub provided you some misleadning context in the email notification, though

File: lldb/source/Plugins/LanguageRuntime/CPlusPlus/CPPLanguageRuntime.cpp (hunk 6 above)
Comment (Michael137) [reply lldb/source/Plugins/LanguageRuntime/CPlusPlus/CPPLanguageRuntime.cpp:91]: Personally I prefer not having that extra `matches_regex` variable. The first time I was reading through it I was trying to figure out if there was more to this loop that I was missing. So if you don't mind changing it to:
```
   for (RegularExpression &r : m_hidden_regex) {
      if (!r.Execute(sc.function->GetNameNoArguments()))
        continue;

      // remaining checks
    }
```
that would be appreciated

**Review by Michael137 — APPROVED** [review]

**Comment by jimingham** [comment]
> @jimingham I agree with everything you wrote. I think you misunderstood the context based on the way GitHub provided you some misleading context in the email notification

Yeah, I should stop trying to read the notifications in the GitHub emails and just use them as a link to the website.  Sigh...

Anyway, the real question was "what does an empty string do in these regex's".  The answer is the one you probably suspected, an empty string means "no regular expression to apply", not ".*".

File: lldb/source/Plugins/LanguageRuntime/CPlusPlus/CPPLanguageRuntime.cpp [hunk 7]
```diff
@@ -90,9 +79,28 @@ class LibCXXFrameRecognizer : public StackFrameRecognizer {
...
+      if (!parent_frame_sp)
+        return {};
+      const auto &parent_sc =
+          parent_frame_sp->GetSymbolContext(lldb::eSymbolContextFunction);
+      if (!parent_sc.function)
+        return {};
+      if (parent_sc.function->GetNameNoArguments().GetStringRef().starts_with(
+              "std::")) {
         return m_hidden_frame;
+      }
```
Comment (Michael137) [inline lldb/source/Plugins/LanguageRuntime/CPlusPlus/CPPLanguageRuntime.cpp:102]:
```suggestion
      if (parent_sc.function->GetNameNoArguments().GetStringRef().starts_with(
              "std::"))
        return m_hidden_frame;
```

**Review by Michael137 — APPROVED** [review]

File: lldb/source/Plugins/LanguageRuntime/CPlusPlus/CPPLanguageRuntime.cpp (hunk 7 above)
Comment (vogelsgesang) [reply lldb/source/Plugins/LanguageRuntime/CPlusPlus/CPPLanguageRuntime.cpp:102]: Argh. There really should be a clang-tidy check or something similar for this...

**Comment by vogelsgesang** [comment]
(I will be waiting for one or two more weeks before merging this, in case there are any additional comments)

**Review by adrian-prantl — APPROVED** [review]
The regexes look sane (and sufficiently restricted) to me, thanks!

**Comment by felipepiovezan** [comment]
Hi @vogelsgesang , the test you added on this PR breaks the LLDB matrix bot, in particular it breaks the 15 & 17 clang tests. Could you add a `skipIf` annotation if this test is only supposed to pass when working with newer versions of libcxx/clang? https://green.lab.llvm.org/job/llvm.org/view/LLDB/job/lldb-cmake-matrix/912/execution/node/107/log/

15 error:
```
/Users/ec2-user/jenkins/workspace/llvm.org/lldb-cmake-matrix/llvm-project/lldb/test/API/lang/cpp/libcxx-internals-recognizer/main.cpp:29:3: error: no member named 'sort' in namespace 'std::ranges'; did you mean 'std::sort'?
  std::ranges::sort(vec.begin(), vec.end(), ranges_sort_less);
  ^~~~~~~~~~~~~~~~~
  std::sort
/Users/ec2-user/jenkins/workspace/llvm.org/lldb-cmake-matrix/clang_1501_build/include/c++/v1/__algorithm/sort.h:693:6: note: 'std::sort' declared here
void sort(_RandomAccessIterator __first, _RandomAccessIterator __last, _Comp __comp) {
     ^
/Users/ec2-user/jenkins/workspace/llvm.org/lldb-cmake-matrix/llvm-project/lldb/test/API/lang/cpp/libcxx-internals-recognizer/main.cpp:32:36: error: no member named 'views' in namespace 'std::ranges'
  for (auto x : vec | std::ranges::views::transform(view_transform)) {
                      ~~~~~~~~~~~~~^
2 errors generated.
make: *** [main.o] Error 1

```

17 error:

```
FAIL: test_frame_recognizer (TestLibcxxInternalsRecognizer.LibCxxInternalsRecognizerTestCase)
   Test that implementation details of libc++ are hidden
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/ec2-user/jenkins/workspace/llvm.org/lldb-cmake-matrix/llvm-project/lldb/test/API/lang/cpp/libcxx-internals-recognizer/TestLibcxxInternalsRecognizer.py", line 60, in test_frame_recognizer
    self.assertIn(
AssertionError: 'ranges::__sort::operator()' not found in 'std::__1::__wrap_iter<int*> std::__1::ranges::__sort::__fn::operator()[abi:ue170006]<std::__1::__wrap_iter<int*>, std::__1::__wrap_iter<int*>, bool (*)(int, int), std::__1::identity>(std::__1::__wrap_iter<int*>, std::__1::__wrap_iter<int*>, bool (*)(int, int), std::__1::identity) const'
Config=x86_64-/Users/ec2-user/jenkins/workspace/llvm.org/lldb-cmake-matrix/clang_1706_build/bin/clang
```

**Comment by vogelsgesang** [comment]
see #114122
